tooling/Microsoft.VisualStudio.Mac.RazorAddin/RazorProjectExtension.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Mac.LanguageServices.Razor.ProjectSystem;
+using MonoDevelop.Ide.Composition;
+using MonoDevelop.Ide.TypeSystem;
+using MonoDevelop.Projects;
+
+namespace Microsoft.VisualStudio.Mac.RazorAddin
+{
+    internal class RazorProjectExtension : ProjectExtension
+    {
+        private readonly object ProjectPropertyLock = new object();
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+
+        public RazorProjectExtension()
+        {
+            _foregroundDispatcher = CompositionManager.GetExportedValue<ForegroundDispatcher>();
+        }
+
+        protected override void OnBoundToSolution()
+        {
+            if (!(Project is DotNetProject dotNetProject))
```

Comment:
This is the entry point for this entire PR so a good part to start at when reviewing.